src/engine/raw_bdb_engine.cc

```diff
@@ -0,0 +1,2058 @@
+  } catch (std::exception& std_exception) {
+    DINGO_LOG(ERROR) << fmt::format("[bdb] std exception, {}.", std_exception.what());
+    status_ = butil::Status(pb::error::ESTD_EXCEPTION, fmt::format("std exception, {}.", std_exception.what()));
+  }
+}
+
+void Iterator::Next() {
+  std::cout << "zhangjie-> enter next" << std::endl;
+  valid_ = false;
+  status_ = butil::Status();
+
+  try {
+    int ret = cursorp_->get(&bdb_key_, &bdb_value_, DB_NEXT);
+    if (ret == 0) {
+      valid_ = true;
+      return;
+    }
+
+    if (ret != DB_NOTFOUND) {
+      DINGO_LOG(ERROR) << fmt::format("[bdb] get failed ret: {}.", ret);
+      status_ = butil::Status(pb::error::EINTERNAL, "internal cursor next error.");
+    }
+  } catch (DbDeadlockException&) {
+    DINGO_LOG(ERROR) << fmt::format("[bdb] writer got DeadLockException, giving up.");
+    status_ = butil::Status(pb::error::EBDB_DEADLOCK, "writer got DeadLockException, giving up.");
+  } catch (DbException& db_exception) {
+    DINGO_LOG(ERROR) << fmt::format("[bdb] db cursor next failed, exception: {}.", db_exception.what());
+    status_ = butil::Status(pb::error::EBDB_EXCEPTION, fmt::format("db cursor next failed, {}.", db_exception.what()));
+  } catch (std::exception& std_exception) {
+    DINGO_LOG(ERROR) << fmt::format("[bdb] std exception, {}.", std_exception.what());
+    status_ = butil::Status(pb::error::ESTD_EXCEPTION, fmt::format("std exception, {}.", std_exception.what()));
+  }
+}
+
+void Iterator::Prev() {
+  valid_ = false;
+  status_ = butil::Status();
+
+  try {
+    int ret = cursorp_->get(&bdb_key_, &bdb_value_, DB_PREV);
+    if (ret == 0) {
+      valid_ = true;
+      return;
+    }
+
+    if (ret != DB_NOTFOUND) {
+      DINGO_LOG(ERROR) << fmt::format("[bdb] get failed ret: {}.", ret);
+      status_ = butil::Status(pb::error::EINTERNAL, "internal cursor prev error.");
+    }
+  } catch (DbDeadlockException&) {
+    DINGO_LOG(ERROR) << fmt::format("[bdb] writer got DeadLockException, giving up.");
+    status_ = butil::Status(pb::error::EBDB_DEADLOCK, "writer got DeadLockException, giving up.");
+  } catch (DbException& db_exception) {
+    DINGO_LOG(ERROR) << fmt::format("[bdb] db cursor prev failed, exception: {}.", db_exception.what());
+    status_ = butil::Status(pb::error::EBDB_EXCEPTION, fmt::format("db cursor prev failed, {}.", db_exception.what()));
+  } catch (std::exception& std_exception) {
+    DINGO_LOG(ERROR) << fmt::format("[bdb] std exception, {}.", std_exception.what());
+    status_ = butil::Status(pb::error::ESTD_EXCEPTION, fmt::format("std exception, {}.", std_exception.what()));
+  }
+}
+
+// Snapshot
+Snapshot::~Snapshot() {
+  if (txn_ != nullptr) {
+    std::cout << "~Snapshot() ======= txn_ is not nullptr" << std::endl;
+    int ret = 0;
+    // commit
+    try {
+      ret = txn_->commit(0);
+      if (ret == 0) {
+        txn_ = nullptr;
```

Comment:
不用，头文件中有下面的注释，源码中的说法：
  // Note: use DbEnv::txn_begin() to get pointers to a DbTxn,
  // and call DbTxn::abort() or DbTxn::commit rather than
  // delete to release them.